### PR TITLE
samples: fix GPIO API usage issues

### PR DESCRIPTION
**Pull request by gmarull**
- Add missing CONFIG_GPIO=y
- Add/remove header if used/not used

A tiny step towards the removal of `CONFIG_GPIO=y` from all boards.

### CODE REVIEW

**Review by fabiobaltieri — DISMISSED** [review]

File: samples/subsys/tracing/prj.conf [hunk 1]
```diff
@@ -5,3 +5,4 @@ CONFIG_MP_MAX_NUM_CPUS=1
 # Debugging
 CONFIG_DEBUG_OPTIMIZATIONS=n
 CONFIG_DEBUG_THREAD_INFO=y
+CONFIG_GPIO=y
```
Comment (kartben) [inline samples/subsys/tracing/prj.conf:8]: drop -- this is only for GPIO tracing AFAICT and enabled in prj_gpio.conf (https://docs.zephyrproject.org/latest/samples/subsys/tracing/README.html#usage-for-gpio-tracing-backend).

Can you maybe double check whatever grep logic you used to ensure it's only for sample actually using gpio API, and in their main.c, that you enable the Kconfig? I suspect this particular sample might not be the only one where this happened.

**Review by kartben — CHANGES_REQUESTED** [review]

File: samples/bluetooth/iso_receive/prj.conf [hunk 2]
```diff
@@ -5,3 +5,4 @@ CONFIG_LOG=y
 CONFIG_BT_DEVICE_NAME="Test ISO Receive"
 
 CONFIG_BT_ISO_MAX_CHAN=2
+CONFIG_GPIO=y
```
Comment (Thalley) [inline samples/bluetooth/iso_receive/prj.conf:8]: This PR is doing something similar to what https://github.com/zephyrproject-rtos/zephyr/pull/81435 attempted, and my review comment is the same:
The GPIO usage in many/most of these samples are optional, i.e. it most cases I am aware of, it checks if there is an LED and if there is, it toggles that. 

Enabling GPIO in the prj.conf file will provide build warnings/errors for boards that does not have that, even though the boards can run the sample without the LED toggle just fine. 

In other words, GPIO is optional in many of these samples, and this is effectively making it mandatory. 

I'm not familiar with all samples, and some may actually require it, but the `iso_receive` and `periodic_sync` samples only use LED if it is there, and if there's no LED there's no GPIO usage. 

@cvinayak I think you've added this for some of these samples. What are your thoughts?

**Review by Thalley — CHANGES_REQUESTED** [review]

File: samples/bluetooth/iso_receive/prj.conf (hunk 2 above)
Comment (gmarull) [reply samples/bluetooth/iso_receive/prj.conf:8]: For such cases, the sample should:

- Have a Kconfig option that selects GPIO if LED exists and enabled in DT
- Or, have an alternate config (e.g. in prj_gpio.conf) that can be used to test with GPIO enabled

Which one is preferred?

Comment (gmarull) [reply samples/bluetooth/iso_receive/prj.conf:8]: > > if a board has a LED defined, build will fail if it doesn't have CONFIG_GPIO=y, so this is wrong.
> 
> Is it possible that `DT_NODE_HAS_STATUS_OKAY(LED0_NODE)` returns `1` without GPIO? :o
> 

Yes, this is just checking DT. A LED node could implement other APIs, in fact, e.g. LED API.

> In which case, the check should be `DT_NODE_HAS_STATUS_OKAY(LED0_NODE) && IS_ENABLED(CONFIG_GPIO)`.

That works for me. I'll update the PR.

Comment (Thalley) [reply samples/bluetooth/iso_receive/prj.conf:8]: > That works for me. I'll update the PR.

Great - But we will need to determine where/how to enable GPIO :)

Comment (gmarull) [reply samples/bluetooth/iso_receive/prj.conf]: please check latest patch set

Comment (Thalley) [reply samples/bluetooth/iso_receive/prj.conf]: Now it's back to only being enable by the board itself, isn't it? So once GPIO is removed from boards, it's not longer automatically enabled for board that support the feature? This seems contrary to what the purpose of this PR is supposed to do

Comment (gmarull) [reply samples/bluetooth/iso_receive/prj.conf]: I've added a sample variant that enables GPIO, isn't that sufficient? If not you need to create an auxiliary Kconfig option that defaults to y if DT has a LED node enabled, selects GPIO and is used as a condition to compile that GPIO branches.

Comment (gmarull) [reply samples/bluetooth/iso_receive/prj.conf]: @Thalley check latest patch set, I think it is better.

Comment (Thalley) [reply samples/bluetooth/iso_receive/prj.conf]: > @Thalley check latest patch set, I think it is better.

Yeah, that looks good to me

Should that be applied to all samples where you've set `CONFIG_GPIO=y`?

Comment (gmarull) [reply samples/bluetooth/iso_receive/prj.conf]: Some samples use GPIO unconditionally, so not always. It's true that we could probably find some common patterns and simplify some samples.

**Review by nordicjm — DISMISSED** [review]

File: samples/boards/espressif/deep_sleep/prj.conf [hunk 3]
```diff
@@ -1 +1,2 @@
 CONFIG_POWEROFF=y
+CONFIG_GPIO=y
```
Comment (Thalley) [inline samples/boards/espressif/deep_sleep/prj.conf:2]: Looks like this seem also uses GPIO optionally. The sample already has 2 Kconfig options, and only if they are enabled is GPIO used as far as I can tell. Should be trivial to add `select GPIO` to those

File: samples/drivers/espi/prj.conf [hunk 4]
```diff
@@ -1,3 +1,4 @@
 CONFIG_ESPI=y
 CONFIG_STDOUT_CONSOLE=y
 CONFIG_PRINTK=y
+CONFIG_GPIO=y
```
Comment (Thalley) [inline samples/drivers/espi/prj.conf:4]: Looks like this sample also optionally uses GPIO

File: samples/drivers/i2s/echo/prj.conf [hunk 5]
```diff
@@ -1 +1,2 @@
 CONFIG_I2S=y
+CONFIG_GPIO=y
```
Comment (Thalley) [inline samples/drivers/i2s/echo/prj.conf:2]: Optional GPIO

File: samples/subsys/display/lvgl/prj.conf [hunk 6]
```diff
@@ -17,3 +17,4 @@ CONFIG_LV_USE_ARC=y
 CONFIG_LV_USE_IMG=y
 CONFIG_LV_USE_MONKEY=y
 CONFIG_LV_FONT_MONTSERRAT_14=y
+CONFIG_GPIO=y
```
Comment (Thalley) [inline samples/subsys/display/lvgl/prj.conf:20]: Optional GPIO

**Review by Thalley — COMMENTED** [review]
Changes LGTM

Some samples seem to optionally support GPIO but still has `GPIO=y`, but those are outside my area, so not sure whether these should be updated. 

Will await with approval a bit

File: samples/boards/espressif/deep_sleep/prj.conf (hunk 3 above)
Comment (gmarull) [reply samples/boards/espressif/deep_sleep/prj.conf]: fixed

File: samples/drivers/i2s/echo/prj.conf (hunk 5 above)
Comment (gmarull) [reply samples/drivers/i2s/echo/prj.conf]: addressed

File: samples/drivers/espi/prj.conf (hunk 4 above)
Comment (gmarull) [reply samples/drivers/espi/prj.conf]: addressed

File: samples/subsys/display/lvgl/prj.conf (hunk 6 above)
Comment (gmarull) [reply samples/subsys/display/lvgl/prj.conf]: addressed

**Review by Thalley — APPROVED** [review]

**Review by nordicjm — APPROVED** [review]

**Review by kartben — APPROVED** [review]

**Review by masz-nordic — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]
